internal/generator/fri/template/fri.go.tmpl

```diff
@@ -0,0 +1,700 @@
+
+	// the salt is binded to the first challenge, to ensure the challenges
+	// are different at each round.
+	err := fs.Bind(xis[0], salt.Marshal())
+	if err != nil {
+		return round{}, err
+	}
+
+	// step 1 : fold the polynomial using the xi
+
+	// evalsAtRound stores the list of the nbSteps polynomial evaluations, each evaluation
+	// corresponds to the evaluation o the folded polynomial at round i.
+	evalsAtRound := make([][]fr.Element, s.nbSteps)
+
+	// evaluate p and sort the result
+	_p := make([]fr.Element, s.domain.Cardinality)
+	copy(_p, p)
+
+	// gInv inverse of the generator of the cyclic group of size the size of the polynomial.
+	// The size of the cyclic group is ρ*s.domainSize, and not s.domainSize.
+	var gInv fr.Element
+	gInv.Set(&s.domain.GeneratorInv)
+
+	for i := 0; i < s.nbSteps; i++ {
+
+		evalsAtRound[i] = sort(_p)
+
+		// compute the root hash, needed to derive xi
+		t := merkletree.New(s.h)
+		for k := 0; k < len(_p); k++ {
+			t.Push(evalsAtRound[i][k].Marshal())
+		}
+		rh := t.Root()
+		err := fs.Bind(xis[i], rh)
+		if err != nil {
+			return res, err
+		}
+
+		// derive the challenge
+		bxi, err := fs.ComputeChallenge(xis[i])
+		if err != nil {
+			return res, err
+		}
+		var xi fr.Element
+		xi.SetBytes(bxi)
+
+		// fold _p, reusing its memory
+		_p = foldPolynomialLagrangeBasis(evalsAtRound[i], gInv, xi)
+
+		// g <- g²
+		gInv.Square(&gInv)
+
+	}
+
+	// last round, provide the evaluation. The fully folded polynomial is of size rho. It should
+	// correspond to the evaluation of a polynomial of degree 1 on ρ points, so those points
+	// are supposed to be on a line.
+	res.evaluation = make([]fr.Element, rho)
+	copy(res.evaluation, _p)
+
+	// step 2: provide the Merkle proofs of the queries
+
+	// derive the verifier queries
+	for i := 0; i < len(res.evaluation); i++ {
+		err := fs.Bind(xis[s.nbSteps], res.evaluation[i].Marshal())
+		if err != nil {
+			return res, err
+		}
+	}
+	binSeed, err := fs.ComputeChallenge(xis[s.nbSteps])
+	if err != nil {
+		return res, err
+	}
+	var bPos, bCardinality big.Int
+	bPos.SetBytes(binSeed)
+	bCardinality.SetUint64(s.domain.Cardinality)
+	bPos.Mod(&bPos, &bCardinality)
+	si := s.deriveQueriesPositions(int(bPos.Uint64()), int(s.domain.Cardinality))
+
+	for i := 0; i < s.nbSteps; i++ {
+
+		// build proofs of queries at s[i]
+		t := merkletree.New(s.h)
+		err := t.SetIndex(uint64(si[i]))
+		if err != nil {
+			return res, err
+		}
+		for k := 0; k < len(evalsAtRound[i]); k++ {
+			t.Push(evalsAtRound[i][k].Marshal())
+		}
+		mr, proofSet, _, numLeaves := t.Prove()
+
+		// c denotes the entry that contains the full Merkle proof. The entry 1-c will
+		// only contain 2 elements, which are the neighbor point, and the hash of the
+		// first point. The remaining of the Merkle path is common to both the original
+		// point and its neighbor.
+		c := si[i] % 2
+		res.interactions[i][c] = partialMerkleProof{mr, proofSet, numLeaves}
+		res.interactions[i][1-c] = partialMerkleProof{
```

Comment:
`1-c` is never `< 0`?

Reply:
no, here c is 0 or 1 here. The c-th entry contains the full Merkle path, the (1-c)-th the partial Merkle path.